src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs

```diff
@@ -95,6 +95,106 @@ public void Kill()
+                */
+                OpenProcessHandle();
+
+                // Kill the process, so that no further children can be created.
+                Kill();
+            }
+            catch
+            {
+                // Made a best-effort attempt which failed (perhaps because the process is already dead), so give up.
+                return;
+            }
+
+            KillChildren(GetChildProcesses());
+        }
+
+        /// <summary>
+        /// Returns all immediate child processes.
+        /// </summary>
+        private IReadOnlyList<Process> GetChildProcesses()
+        {
+            var childProcesses = new List<Process>();
+
+            foreach (Process possibleChildProcess in GetProcesses())
+            {
+                var keep = false;
+
+                try
+                {
+                    try
+                    {
+                        /*
+                           Force the process object to hold a handle to the process. Ensures that if the process dies while we're working
+                           with it, it won't be reused. This way, any process we pass back is guaranteed to be an actual child, not a 
+                           reference to a new process that happens to have the same id as a deceased child.
+                        */
+                        possibleChildProcess.OpenProcessHandle();
+                    }
+                    catch
+                    {
+                        // Made a best-effort attempt which failed (perhaps because the process is already dead).
+                        continue;
+                    }
+
+                    if (IsParentOf(possibleChildProcess))
+                    {
+                        childProcesses.Add(possibleChildProcess);
+                        keep = true;
+                    }
+                }
+                finally
+                {
+                    if (!keep)
+                    {
+                        possibleChildProcess.Dispose();
+                    }
+                }
+            }
+
+            return childProcesses;
+        }
+
+        /// <summary>Checks whether the argument is a direct child of this process.</summary>
+        /// <remarks>
+        /// A child process is a process which has this process's id as its parent process id and which started after this process did.
+        /// </remakrs>
+        private bool IsParentOf(Process possibleChildProcess) =>
+            StartTime < possibleChildProcess.StartTime
+            && possibleChildProcess.TryGetParentProcessId(out var parentProcessId)
+            && Id == parentProcessId;
+
+        /// <summary>
+        /// Attempts to get the process's parent process id.
+        /// </summary>
+        private bool TryGetParentProcessId(out int? parentProcessId)
+        {
+            // UNDONE: NtQueryInformationProcess will fail if we are not elevated and other process is. Advice is to change to use ToolHelp32 API's
```

Comment:
These comments copied verbatim from msbuild's NativeMethodsShared. Unsure whether its still relevant. NativeMethodsShared would call this on an arbitrary process while now this logic is running in the context of the process being inquired about.